### PR TITLE
ismp-testsuite update

### PR DESCRIPTION
**Pull request by gr4yha7**

### CODE REVIEW

File: modules/ismp/testsuite/src/lib.rs [hunk 1]
```diff
@@ -402,7 +434,188 @@ pub fn check_request_source_and_destination() {}
 
 /// This should check that if a proxy isn't configured, responses are not valid if they don't come
 /// from the state machine claimed in the proof
-pub fn check_response_source() {}
+pub fn check_response_source<H, D>(host: &H, dispatcher: &D) -> Result<(), &'static str>
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:437]: Test is in the wrong direction.
First assert that host.allowed_proxy().is_none().
Then try to handle a request and response whose destinatin chain != host.state_machine()
Assert that the request or response is ignored.

File: modules/ismp/testsuite/src/lib.rs [hunk 2]
```diff
@@ -402,7 +434,188 @@ pub fn check_request_source_and_destination() {}
...
+{
+    let proxy_state_commitment = setup_mock_proxy_client(host);
+    // Set the previous update time
+    let challenge_period = host.challenge_period(mock_proxy_consensus_state_id()).unwrap();
+    let previous_update_time = host.timestamp() - (challenge_period * 2);
+    host.store_consensus_update_time(mock_proxy_consensus_state_id(), previous_update_time)
+        .unwrap();
+    host.store_state_machine_update_time(proxy_state_commitment.height, previous_update_time)
+        .unwrap();
+
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:550]: Assert that a proxy is configured and the request destination != host state machine id.

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: modules/ismp/testsuite/src/lib.rs [hunk 3]
```diff
@@ -402,7 +429,128 @@ pub fn check_request_source_and_destination() {}
...
+        from: vec![0u8; 32],
+        to: vec![0u8; 32],
+        timeout_timestamp: 0,
+        data: vec![0u8; 64],
+        gas_limit: 0,
+    };
+    let response =
+        PostResponse { post: post.clone(), response: vec![], timeout_timestamp: 0, gas_limit: 0 };
+    dispatcher
+        .dispatch_response(response.clone(), [0; 32].into(), 0u32.into())
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:485]: We want to test handling of a post response message not dispatching a post response. The test should try to handle a post response where the source of the response does not match the proof source.

File: modules/ismp/testsuite/src/lib.rs [hunk 4]
```diff
@@ -402,7 +429,128 @@ pub fn check_request_source_and_destination() {}
...
+    let response_message = Message::Response(ResponseMessage {
+        datagram: RequestResponse::Response(vec![Response::Post(response.clone())]),
+        proof: Proof { height: intermediate_state.height, proof: vec![] },
+        signer: vec![],
+    });
+
+    handle_incoming_message(host, response_message).unwrap();
+    assert_ne!(response.dest_chain(), host.host_state_machine());
+    // Assert that response was acknowledged
+    // assert!(host.response_receipt(&Response::Post(response)).is_some());
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:553]: You still need these assertions

**Comment by gr4yha7** [comment]
@Wizdave97 PR updated.

File: modules/ismp/testsuite/src/lib.rs [hunk 5]
```diff
@@ -402,7 +429,116 @@ pub fn check_request_source_and_destination() {}
...
+        nonce: 0,
+        from: vec![0u8; 32],
+        to: vec![0u8; 32],
+        timeout_timestamp: 0,
+        data: vec![0u8; 64],
+        gas_limit: 0,
+    };
+    let response =
+        PostResponse { post: post.clone(), response: vec![], timeout_timestamp: 0, gas_limit: 0 };
+    // Response message handling check
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:478]: Add an assertion that proof state id != response.source_chain

**Comment by gr4yha7** [comment]
@Wizdave97 PR updated.

File: modules/ismp/testsuite/src/mocks.rs [hunk 6]
```diff
@@ -372,15 +379,22 @@ impl IsmpHost for Host {
     }
 
     fn allowed_proxy(&self) -> Option<StateMachine> {
-        Some(StateMachine::Kusama(2000))
+        self.consensus_state(*b"prox").map(|_| StateMachine::Kusama(2000)).ok()
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/mocks.rs:382]: What's the purpose of this? Unnecessary

File: modules/ismp/testsuite/src/mocks.rs [hunk 7]
```diff
@@ -410,13 +424,37 @@ impl IsmpModule for MockModule {
     }
 }
 
+#[derive(Default)]
+pub struct MockProxyModule;
+
+impl IsmpModule for MockProxyModule {
+    fn on_accept(&self, _request: Post) -> Result<(), Error> {
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/mocks.rs:431]: Your proxy module should re-dispatch the requests and responses

File: modules/ismp/testsuite/src/lib.rs [hunk 8]
```diff
@@ -624,8 +624,122 @@ pub fn check_request_source_and_destination() {}
...
+        proof: Proof { height: intermediate_state.height, proof: vec![] },
+        signer: vec![],
+    });
+
+    handle_incoming_message(host, response_message).unwrap();
+
+    assert_ne!(response.dest_chain(), host.host_state_machine());
+    // Assert that response was acknowledged
+    assert!(host.response_receipt(&Response::Post(response)).is_some());
+
```
Comment (Wizdave97) [inline modules/ismp/testsuite/src/lib.rs:743]: Test should assert that the proxied requests and responses were re-dispatched.

**Review by Wizdave97 — CHANGES_REQUESTED** [review]

File: modules/ismp/testsuite/src/mocks.rs (hunk 6 above)
Comment (gr4yha7) [reply modules/ismp/testsuite/src/mocks.rs:376]: When a proxy state machine is not configured, the assertion in the following test snippet fails (which is undesirable):
```rust
pub fn check_response_source<H: IsmpHost>(host: &H) -> Result<(), &'static str> {
    ...
    assert!(host.allowed_proxy().is_none());
```

**Comment by Wizdave97** [comment]
I'll add these commits to my next PR.

**Pull request closed without merging**